google/resumable_media/_helpers.py

```diff
@@ -114,6 +140,81 @@ def require_status_code(response, status_codes, callback=do_nothing):
     return status_code
 
 
+def calculate_retry_wait(num_retries):
+    """Calculate the amount of time to wait before a retry attempt.
+
+    Wait time grows exponentially with the number of attempts, until
+    it hits :attr:`MAX_SLEEP`.
+
+    A random amount of jitter (between 0 and 1 seconds) is added to spread out
+    retry attempts from different clients.
+
+    Args:
+        num_retries (int): The number of retries already attempted. If
+            no retries have been attempted, should return 1 second with
+            some jitter.
+
+    Returns:
+        float: The number of seconds to wait before retrying (with a random
+        amount of jitter between 0 and 1 seconds added).
+    """
+    if num_retries < MAX_SLEEP_EXPONENT:
+        wait_time = 2**num_retries
+    else:
+        wait_time = MAX_SLEEP
+
+    jitter_ms = random.randint(0, 1000)
+    return wait_time + 0.001 * jitter_ms
+
+
+def wait_and_retry(func, predicate):
+    """Attempts to retry a call to ``func`` until success.
+
+    Success is determined by ``predicate(result)`` being true.
+
+    Will retry until :attr:`MAX_CUMULATIVE_RETRY` seconds of wait time
+    have accrued. Uses :func:`calculate_retry_wait` to double the
+    wait time (with jitter) after each attempt.
+
+    Args:
+        func (Callable): A callable that takes no arguments and produces
+            a result which will be checked by ``predicate``.
+        predicate (Callable): Determines if the result is valid.
+
+    Returns:
+        object: The return value of ``func``.
+    """
+    result = func()
+    if predicate(result):
+        return result
+
+    total_sleep = 0.0
+    num_retries = 0
+    while total_sleep < MAX_CUMULATIVE_RETRY:
+        wait_time = calculate_retry_wait(num_retries)
+        num_retries += 1
+        total_sleep += wait_time
+        time.sleep(wait_time)
+        result = func()
+        if predicate(result):
+            return result
+
+    return result
+
+
+def not_retryable_predicate(response):
+    """Determines if a ``response`` is retryable.
+
+    Args:
+        object: The return value of ``transport.request()``.
```

Comment:
The idea is that `transport` can be anything, which is why I'm strongly considering making `google.resumable_media.FOO` subpackages so it is **ALWAYS** explicit